src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,743 @@
+  where
+    opts :: ParserInfo PactImportConfig
+    opts = O.info (parser <**> O.helper) (O.fullDesc <> O.progDesc helpText)
+
+    helpText :: String
+    helpText = unlines
+      [ "Compare the grand hash of a Pact database to an expected value."
+      , "If the hash matches, optionally import the database into a target directory"
+      , "and delete any newer state not included in the hash."
+      ]
+
+    parser :: Parser PactImportConfig
+    parser = PactImportConfig
+      <$> O.strOption
+            (O.long "source-database-dir"
+             <> O.short 's'
+             <> O.metavar "PACT DBDIR"
+             <> O.help "Source Pact database directory, which you wish to verify"
+            )
+      <*> optional (O.strOption
+            (O.long "target-database-dir"
+             <> O.short 't'
+             <> O.metavar "PACT DBDIR"
+             <> O.help "Target directory to copy the verified Pact database into"
+            ))
+      <*> rocksParser
+      <*> cwvParser
+
+versionFromText :: (HasCallStack) => Text -> ChainwebVersion
+versionFromText t = case fromText @ChainwebVersionName t of
+  Just a -> lookupVersionByName a
+  Nothing -> error $ "Invalid chainweb version name: " ++ Text.unpack t
+
+checkPactDbsExist :: FilePath -> [ChainId] -> IO ()
+checkPactDbsExist dbDir cids = pooledFor_ cids $ \cid -> do
+  e <- doesFileExist (chainwebDbFilePath cid dbDir)
+  when (not e) $ do
+    error $ "Pact database doesn't exist for expected chain id " <> Text.unpack (chainIdToText cid)
+
+chainwebDbFilePath :: ChainId -> FilePath -> FilePath
+chainwebDbFilePath cid dbDir =
+  let fileName = mconcat
+        [ "pact-v1-chain-"
+        , Text.unpack (chainIdToText cid)
+        , ".sqlite"
+        ]
+  in dbDir </> fileName
+
+cwvParser :: Parser ChainwebVersion
+cwvParser = fmap versionFromText $ O.strOption
+  (O.long "graph-version"
+    <> O.short 'v'
+    <> O.metavar "CHAINWEB_VERSION"
+    <> O.help "Chainweb version for graph. Only needed for non-standard graphs."
+    <> O.value (toText (_versionName mainnet))
+    <> O.showDefault
+  )
+
+rocksParser :: Parser FilePath
+rocksParser = O.strOption
+  (O.long "rocksdb-dir"
+   <> O.metavar "ROCKSDB DIR"
+   <> O.help "Path to RocksDB directory"
+  )
+
+targetsParser :: Parser BlockHeightTargets
+targetsParser =
+  let
+    makeTargeted = \case
+      [] -> LatestAll
+      ts -> TargetAll (Set.fromList ts)
+    p = fmap makeTargeted $ many $ fmap BlockHeight $ O.option O.auto
+          (O.long "target-blockheight"
+            <> O.short 'b'
+            <> O.metavar "BLOCKHEIGHT"
+            <> O.help "BlockHeight to verify."
+          )
+  in
+  p <|> pure LatestAll
+
+hex :: ByteString -> Text
+hex = Text.decodeUtf8 . Base16.encode
+
+pooledFor :: (Traversable t) => t a -> (a -> IO b) -> IO (t b)
+pooledFor = pooledForConcurrentlyN 4
+
+pooledFor_ :: (Foldable t) => t a -> (a -> IO b) -> IO ()
+pooledFor_ = pooledForConcurrentlyN_ 4
```

Comment:
let's replace `4` with a named constant?

Reply:
Can't we use `getNumCapabilities` somehow here? Doesn't `pooledForConcurrently` do that already?

Reply:
This is a holdover from WIP stage.